copy-paster

szymonkaliski opened the pull request:
## Roadmap

- [x] dumb components for template list and template editor
- [x] templating language research
- [x] wiring dumb components into the extension with mocked data
- [x] interfacing with clipboard
- [x] connecting UI components to the store / storage class (@ShishKabab)
  - [x] storexhub CRUD
  - [x] connect to UI
    - [x] create new template
    - [x] edit template
    - [x] remove template
- [x] UI feedback for clicking on template to copy into clipboard

## Open Questions/Issues

- [x] ~~where/how to wire `onClickOutside`? - currently wired nowhere, hence clicking outside does nothing~~ - figured this out
- [ ] `monospaced` font for formula editor

---

Initial PR includes dumb TSX components and storybook stories for them.

Some thoughts/questions:
1. "remove" button is not on the template row - not in figma also - but might be a good addition (just like we have "delete" button for the search list results) - or should we "delete" only through "edit" modal?
2. switched save/cancel button order in editor to match the order when editing an annotation
3. not sure about the combination of `styled-components` and `require(*.css)` + `className={}` - are we leaning in one direction, or is it ok to use both?
4. template code text area should probably use monospaced font - do we have one?

Additionally - some initial thoughts on templating language: https://gist.github.com/szymonkaliski/8c9297c04a9a3630a1cf412278d98184 - I'm leaning towards handlebars, since they are quite easy to use, and support escaping (and not escaping).

blackforestboi commented:
Sweet! Great progress.

Will note down answers to everything I can answer. The rest @ShishKabab & @cdharris can answer. 

> "remove" button is not on the template row - not in figma also - but might be a good addition (just like we have "delete" button for the search list results) - or should we "delete" only through "edit" modal?

I think it would be good to add it to the edit modal indeed. 

> switched save/cancel button order in editor to match the order when editing an annotation

Good call

> not sure about the combination of styled-components and require(*.css) + className={} - are we leaning in one direction, or is it ok to use both?

If I am correct we are switching to styled components

> template code text area should probably use monospaced font - do we have one?

We don't have a font specifically for that no. 


A side note for the PR content: It would be great to have a checklist in the OP with all the work that has been done and is foreseeable to still do. This way we can keep track of the progress.

szymonkaliski commented:
Thanks for the comments folks!

> > "remove" button is not on the template row - not in figma also - but might be a good addition (just like we have "delete" button for the search list results) - or should we "delete" only through "edit" modal?
>
> I think it would be good to add it to the edit modal indeed.

I think I was unclear - in the figma file the "delete" is in the edit modal, but not on the template list:

![Screenshot 2020-05-28 at 12 41 19](https://user-images.githubusercontent.com/1040420/83131723-94ee8700-a0e0-11ea-9cdc-8c39b55c8c38.png)
![Screenshot 2020-05-28 at 12 41 23](https://user-images.githubusercontent.com/1040420/83131727-96b84a80-a0e0-11ea-972c-5ba141845bd9.png)

> It would be great to have a checklist in the OP with all the work that has been done and is foreseeable to still do. This way we can keep track of the progress.

On it!

> About styled-components, we want to fully switch to it, so it'd be great if you could port over that CSS file.

So the thing is, I'm using the `src/common-ui/components/TextInputControlled.tsx` for text inputs, and it requires `className` for styling - not sure if it makes sense right now to try rewriting it? Or building ad-hoc text inputs for the template editor?

> Nitpick: Let's not have the TemplateRow props inherit Template, but have a template property to keep things consistent and more understandable (you already do this in this other components like TemplateEditor.)

Sure yeah no problem, I wasn't sure about this myself :)

> Looks good so far! Next step I'd say would be to integrate the UI logic. I think it'd be the best if as the first step you just pass in the external dependencies into the top-level component (TemplateList?) and just mock that thing for now.

Yup, sure, I think even before that I want to add the icon to the list (already done) and trigger popup view from it - then the mocks you mentioned - let me know if that makes sense?

blackforestboi commented:
> I think I was unclear - in the figma file the "delete" is in the edit modal, but not on the template list:

Not sure whats better for now but seems like leaving it in the modal itself is cleaner in regards to the list (adding 4 buttons seems a bit much?)

ShishKabab commented:
> Yup, sure, I think even before that I want to add the icon to the list (already done) and trigger popup view from it - then the mocks you mentioned - let me know if that makes sense?

Yeah, all the UI work you can, go ahead!

Also, I've talked a bit with Oliver and thought it may be nice if you could also do some storage work if you want. If so, I can just prepare some boilerplate that you can build upon. You'd basically have to implement something like `src/custom-lists/background` with a BackgroundModule, StorageModule and tests surrounding that. If you'd be up for some Storex stuff, I think that'd be a very nice start  :)

ShishKabab commented:
> I'm using the src/common-ui/components/TextInputControlled.tsx for text inputs, and it requires className for styling

Ah, didn't see that scanning through the code. Yeah, then it makes sense to keep the CSS in an external file for now. The alternative is [global styles](https://styled-components.com/docs/api#createglobalstyle) but it doesn't seem very attractive.

szymonkaliski commented:
Current status:

![2020-05-28 17-12-47 2020-05-28 17_16_22](https://user-images.githubusercontent.com/1040420/83160031-f7f31480-a106-11ea-80a7-32981e0d34af.gif)

- issues:
  - tooltip overflows inside of the template list in popup - what can we do about this?

- questions:
  - where to wire `onClickOutside`? - currently wired nowhere, hence clicking outside does nothing

- next (in order of priorities):
  - [ ] interfacing with pasteboard + handlebar compilation
  - [ ] wire to StorexHub (with help of @ShishKabab)
  - [ ] `monospaced` font (should we pick one now? maybe system fonts are ok for MVP?)

ShishKabab commented:
To implement the storage, you'll need to:

* Copy a background module (like `src/custom-lists/background` to `src/copy-paster/background`) and add it to the background module setup (`src/background-script/setup.ts`) to both the creation function and the `BackgroundModules` interface.
* In there, create a `storage.ts` that you can adapt from `src/custom-lists/storage.ts`

In the storage module, you can see a few things going on. One is the `collections`, which for now you can put straight in there instead of in `external/@worldbrain/memex-common`. They should be fairly straightforward to define. Every `collection` is basically a database table, containing the field types and its indices.

Then you see `operations`, which I really should've call `operationTemplates`. Here you say which kind of database operation you're going to use in the methods of this class, so we have an overview of all the interactions with the database, so we can easily spot missing indices and can have an idea of how the data is interacted with. These operations can later be used in the methods with `this.operation('template name', { <placeholders> })`.

It's all pretty easy to understand if you read through the `CustomListsStorage`. Also in there, you can find `storage.test.ts`, which gives you an idea how to automatically test this. I recommend developing the entire thing TDD-style.

As for the remote calls connecting the UI with the background, I think you already have a fairly solid grip on that, right?

ShishKabab commented:
Just looked at your template language tests. Would be great if the user could control the escaping, which should be possible with handlebars. (Making helpers like `json_escape`, `html_escape`, etc. Your test don't account yet for rendering a list of things in one go, which introduces things like trailing comma's in JSON (which parsers really care about.) However, Handlebars should be able to handle that if we introduce a helper to see if it's the last item we're rendering.

szymonkaliski commented on src/options/options.jsx:
```diff
@@ -18,6 +18,8 @@ const ReduxDevTools =
 
 const store = configureStore({ ReduxDevTools })
 
+window.store = store
```
yeah, commited by accident - for some reason I had problems running redux / react tooling against the extension's dashboard - is there some magic way to get them going, or to debug/browse through the store state?

ShishKabab replied:
I think the devtools cannot access extension pages. Sucks quite badly... Also, while Redux has `store.subscribe()`, it doesn't tell you what changed :facepalm:

szymonkaliski replied:
yeah, I just do `window.store.getState()` and explore as needed - maybe we should keep this when `NODE_ENV != "production"`?

szymonkaliski commented:
Current status - got clipboard working:

![clip 2020-06-03 11_03_22](https://user-images.githubusercontent.com/1040420/83618059-39157980-a58a-11ea-9f39-d7f1df409a08.gif)

It looks like we'll need some feedback to the user that the "copy" happened - the button doesn't even have the "clicked" state - I don't think anything in the extension has that state handled visually. I could add grey background when hovered, but that doesn't solve everything - any thoughts @blackforestboi ?

Also, switched from handlebars to mustache (same syntax though) - handlebars implementation uses `eval` which doesn't play nice with chrome extensions, moustache is `eval`-free.

> Would be great if the user could control the escaping, which should be possible with handlebars. (Making helpers like json_escape, html_escape, etc.) Your test don't account yet for rendering a list of things in one go, which introduces things like trailing comma's in JSON (which parsers really care about.) However, Handlebars should be able to handle that if we introduce a helper to see if it's the last item we're rendering.

All good points! First thing - I think this PR/scope is only about copying one item at a time - I agree that copying multiple items has an issue with trailing commas, let's maybe keep this in mind for v2?

As for escaping - I think I already touched on that in the gist, but basically using `{{thing}}` escapes `thing`, and `{{{thing}}}` does not - https://github.com/janl/mustache.js#variables

> All variables are HTML-escaped by default. If you want to render unescaped HTML, use the triple mustache: `{{{name}}}`. You can also use `&` to unescape a variable.

Next step for me is the storage implementation. 

(Top comment also updated)

blackforestboi commented:
> I could add grey background when hovered, but that doesn't solve everything - any thoughts @blackforestboi ?

Can you set a state for the icon "copied" and add a random icon? (will change it in the review/testing process). It should go back to normal after 5 seconds or so. 
Do you think that would be good UX?

> All good points! First thing - I think this PR/scope is only about copying one item at a time - I agree that copying multiple items has an issue with trailing commas, let's maybe keep this in mind for v2?

Yeah lets focus now on just single items in the V1.

blackforestboi commented:
Another question: 
The "markdown" option is also just a template like the HTML one, right? We don't make opionionated suggestions there yet?

As of the designs for the "copy state" let me make some mocks on figma and post them here in a bit!

szymonkaliski commented:
> The "markdown" option is also just a template like the HTML one, right? We don't make opionionated suggestions there yet?

Yes! Mocked here: https://github.com/WorldBrain/Memex/pull/1042/files#diff-5755c5a56ced4d9931dc7ab12d56f35bR252 - I think we should provide some "default templates" on start, but user can edit/remove them as needed.

> Can you set a state for the icon "copied" and add a random icon? (will change it in the review/testing process). It should go back to normal after 5 seconds or so.
> Do you think that would be good UX?

Hm, we can do that, but what would happen when you click the template the second time - no feedback then, as the icon is already there? I was thinking of tooltip showing "Copied!" for a few seconds, and disappearing, so if you copy the same elementy multiple times, it still shows up - what do you think?

blackforestboi commented:
> I think we should provide some "default templates" on start, but user can edit/remove them as needed.

Yeah we can provide them, but they should not be hard coded in a way the user can't remove or edit them indeed. 

 >Hm, we can do that, but what would happen when you click the template the second time - no feedback then, as the icon is already there? 

we can just make the "copied to clipboard" text reappear again - have some simple animation (move up from bottom). The copy button could stay there all the time, so the user could if they like click multiple times?

szymonkaliski commented:
@ShishKabab when you have few minutes, could you review the `background/*` files if I understood everything correctly? Tests pass, did the implementation TDD style, and only thing left over here is to wire this back to UI through `remoteFunction()`.

ShishKabab commented on src/overview/copy-paster/background/index.ts:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    async createTemplate({ title, code, isFavourite = false }) {
```
Could you please add full types to these signatures? What I normally do to prevent redundancy is write it as:
```
createTemplate(params: { title: string, code: string, isFavourite?: boolean }) 
```

ShishKabab commented on src/overview/copy-paster/background/storage.test.ts:
```diff
@@ -0,0 +1,92 @@
+
+    const copyPaster: CopyPasterBackground = backgroundModules.copyPaster
+
+    registerModuleMapCollections(storageManager.registry, {
```
This should be done automatically, so no need to do it yourself.

ShishKabab commented on src/overview/copy-paster/background/storage.test.ts:
```diff
@@ -0,0 +1,92 @@
+}
+
+describe('create template', () => {
+    test('should be able to create template', async () => {
```
Nitpick: don't forget articles, also in other tests, like 'should be able to create **A** template' or 'create templates'. I know articles are not there in Slavic languages (I speak Czech myself  :)  )

ShishKabab commented on src/overview/copy-paster/background/index.ts:
```diff
@@ -0,0 +1,62 @@
+            storageManager: options.storageManager,
+        })
+
+        this.remoteFunctions = {
```
There's no types for these yet.

blackforestboi commented:
I wanna write the instructions for this but need to know how we identify each field in the syntax of our template language. 
I'd prefer if the user is capable of using every single field in the DB available to the page/annotation object, and use a syntax that matches the database keys/names of those fields.

szymonkaliski commented:
Hey @ShishKabab - I think I fixed all your comments. I started wiring "new template" to the store, the way it currently works is:
1. when clicking "new template" a `tmpCopyPasterTemplate` is create in `state`
2. it's opened in the `TemplateEditor` and any changes are stored in `state`
3. when clicking "save" the `tmpCopyPasterTemplate` is pushed to `runInBackground`'s `createTemplate` - which fails with an error I'm having hard time tracking:

```
"Failed to execute 'put' on 'IDBObjectStore': Evaluating the object store's key path yielded a value that is not a valid key.
DataError: Failed to execute 'put' on 'IDBObjectStore': Evaluating the object store's key path yielded a value that is not a valid key."
```

strangely the tests on `createTemplate` pass, so I must be doing something stupid here, I'd love some guidance!

Good news is - once this is sorted, I'm super close to connecting all the other methods and having MVP version ready to test!

ShishKabab commented:
Looks like it's having some trouble with the primary key. Anyway, auto incremented integers don't play very well with the sync atm, so could you change the `id` field to a string, remove `autoInc` and put a random string in there? Might resolve the issue.

szymonkaliski commented:
Yeah, `autoInc` was the problem, thanks! Will generate ids from `Date.now()` for now (which I saw few times in the codebase already). Thanks :)

blackforestboi commented:
Gave it a go and here are a few notes to improvements: 
- I can't edit a template. Text input is not recognised
- I can't delete a template
- As soon as there is a `{ }`I can't write inside the braces. Only the first letter is recognised, then the cursor jumps out of the braces
- I urgently need a list of all the fields that can be used and the exact titles to write the docs.

szymonkaliski commented:
Hey all, I think everything is here!

![2020-06-08 15-52-20 2020-06-08 15_53_50](https://user-images.githubusercontent.com/1040420/84038578-673af500-a9a0-11ea-86f8-cc751d7444c9.gif)

> I urgently need a list of all the fields that can be used and the exact titles to write the docs.

- so far I'm passing through `title: string`, `url: string` and `tags: string[]` - I can add whatever, just let me know what else makes sense?

> remove the "bookmark" template button. It's ok to have the copy paster for now with a click more to select a template. (reducing scope to launch with less feature, but working)

- damn, I'm not sure if I understand this - could you rephrase / add a screenshot here?

blackforestboi commented:
Couple of things that still need some work: 

-[ ] Adding the copy button to the annotation cards
-[ ] adding "highlight.text" and "highlight.note"  as copiable items. Those would only be pulled when copying an annotation card, however should **not** throw an error if a field is not available (e.g. because of clicking on the wrong template)
-[ ] fixing that you can't click inside the text and edit from there. it always jumps to the end of the text
-[x] removing the "star" because it does only have a purpose if there is no fast way of copying (e.g. by only clicking on the copy button in the result icon, but that is reserved for opening the copy/paster)
- [ ] adding analytics to track: 
    - copy paster used (by click on A templates copy button, not the copy/paste button in the result item)
    - template created

@szymonkaliski since you are likely have tons of stuff on your plate this week for other client work (that pays your bills) it may be good if @cdharris takes over here. 
We found that getting the copy paster out is critical, so we wanna focus on having a version out soon. 

Thanks so much so far for this contribution, it's going to be an amazing feature!

blackforestboi commented on src/overview/copy-paster/background/storage.test.ts:
```diff
@@ -0,0 +1,76 @@
+    return { copyPaster }
+}
+
+describe('Tempaltes', () => {
```
Seems like a typo here

szymonkaliski commented:
should we close this one since #1059 got merged?

blackforestboi commented:
Yeah makes sense. 

Thanks so much @szymonkaliski for contributing so significantly to this feature. Amazing work!
